packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js

```diff
@@ -13,101 +13,522 @@
+    WHERE product_quantity IS NOT NULL
+    AND product_quantity_unit IN ${rawWeightUnits}
+  `);
+
+  // Product not required for pest control task
+  await knex.schema.alterTable('pest_control_task', (table) => {
+    table.check(
+      '(weight IS NULL AND volume IS NOT NULL) OR (volume IS NULL AND weight IS NOT NULL) OR (weight IS NULL AND volume IS NULL)',
+      [],
+      'volume_or_weight_check',
+    );
+    table.check(
+      '(weight IS NULL AND weight_unit IS NULL) OR (weight IS NOT NULL AND weight_unit IS NOT NULL)',
+      [],
+      'weight_unit_check',
+    );
+    table.check(
+      '(volume IS NULL AND volume_unit IS NULL) OR (volume IS NOT NULL AND volume_unit IS NOT NULL)',
+      [],
+      'volume_unit_check',
+    );
   });
 
-  await knex.raw(`
-      UPDATE soil_amendment_task
-      SET temp_purpose = purpose[1]
-    `);
+  // Alter cleaning task for volume and weight
+  await knex.schema.alterTable('cleaning_task', (table) => {
+    table.dropColumn('product_quantity');
+    table.dropColumn('product_quantity_unit');
+  });
 
-  await knex.schema.alterTable('soil_amendment_task', (table) => {
-    table.dropColumn('purpose');
-    table.renameColumn('temp_purpose', 'purpose');
+  // Alter cleaning task for volume and weight
+  await knex.schema.alterTable('pest_control_task', (table) => {
+    table.dropColumn('product_quantity');
+    table.dropColumn('product_quantity_unit');
+  });
+
+  // Alter product to add product state type id
+  // Removing npk to add new table instead no attempt to keep values as it is currently unused
+  await knex.schema.alterTable('product', (table) => {
+    table.dropChecks(['npk_unit_check', 'npk_percent_check']);
+    table.dropColumn('n');
+    table.dropColumn('p');
+    table.dropColumn('k');
+    table.dropColumn('npk_unit');
   });
 
-  await knex.schema.raw('DROP TYPE purpose_enum');
+  //Add permissions
+  // Use task or product permissions as needed
+};
+
+export const down = async function (knex) {
+  // No prod data unreleased feature - destroys data
+  await knex.schema.alterTable('product', (table) => {
+    table.decimal('n');
+    table.decimal('p');
+    table.decimal('k');
+    table.enu('npk_unit', ['ratio', 'percent']);
+    table.check(
+      '(COALESCE(n, p, k) IS NULL AND npk_unit IS NULL) OR (COALESCE(n, p, k) IS NOT NULL AND npk_unit IS NOT NULL)',
+      [],
+      'npk_unit_check',
+    );
+    table.check(
+      "npk_unit != 'percent' OR (npk_unit = 'percent' AND (n + p + k) <= 100)",
+      [],
+      'npk_percent_check',
+    );
+  });
 
   // Reverse product data migration
   await knex.schema.alterTable('soil_amendment_task', (table) => {
     table.decimal('product_quantity', 36, 12);
-    table.enu('product_quantity_unit', productQuantityUnits);
+    table.text('product_quantity_unit');
     table.integer('product_id').references('product_id').inTable('product');
+    table.string('other_purpose');
+    table.text('purpose');
```

Comment:
Same. Purpose is an enum on the current tables.

Reply:
Same I was seeing them as text will double check beta see if i messed up my local or something

Reply:
Purpose is also currently text as I see it: ![image](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/c5039820-5a16-4c4d-a38e-5ed21aa567c9)